Fix route manager

AAlon opened the pull request:
## WIP for experimentation - not for merging - do not review

This combines https://github.com/aws-robotics/aws-robomaker-sample-application-cloudwatch/pull/37 with a fix to route_manager.py.

With these changes you can launch the app and have the robot start moving, but the navigation still fails.

```bash
source /opt/ros/dashing/setup.bash && source /usr/share/gazebo/setup.sh 
source install/local_setup.bash && export TURTLEBOT3_MODEL=waffle_pi
ros2 launch cloudwatch_simulation bookstore_turtlebot_navigation.launch.py gui:=true use_sim_time:=true
```

Logs:
```
[lifecycle_manager-5] [INFO] [lifecycle_manager]: The system is active
[route_manager.py-13] [INFO] [route_manager]: Route manager initialized with 12 goals in random mode
[route_manager.py-13] [INFO] [route_manager]: Route mode is 'random', getting next goal
[route_manager.py-13] [INFO] [route_manager]: Sending target goal: nav2_msgs.action.NavigateToPose_Goal(pose=geometry_msgs.msg.PoseStamped(header=std_msgs.msg.Header(stamp=builtin_interfaces.msg.Time(sec=1568420311, nanosec=666944459), frame_id='map'), pose=geometry_msgs.msg.Pose(position=geometry_msgs.msg.Point(x=5.10259008408, y=0.883781552315, z=0.0), orientation=geometry_msgs.msg.Quaternion(x=0.0, y=0.0, z=-0.455630867938, w=0.890168811059))))
[route_manager.py-13] [INFO] [route_manager]: Goal accepted :)

[bt_navigator-12] [INFO] [bt_navigator]: Begin navigating from current location to (5.10, 0.88)
[bt_navigator-12] [WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
[bt_navigator-12] [ERROR] [getCurrentPose]: No Transform available Error looking up robot pose: "map" passed to lookupTransform argument target_frame does not exist. 
[bt_navigator-12] 
[dwb_controller-9] [INFO] [dwb_controller]: Received a goal, begin following path
[bt_navigator-12] [ERROR] [getCurrentPose]: Extrapolation Error looking up robot pose: Lookup would require extrapolation into the past.  Requested time 7.53900 but the earliest data is at time 7.80200, when looking up transform from frame [base_link] to frame [map]
[bt_navigator-12] 
[bt_navigator-12] [ERROR] [getCurrentPose]: Extrapolation Error looking up robot pose: Lookup would require extrapolation into the past.  Requested time 7.57300 but the earliest data is at time 7.80200, when looking up transform from frame [base_link] to frame [map]
[bt_navigator-12] 
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[world_model-8] [INFO] [global_costmap.global_costmap_rclcpp_node]: [signalFailure] Drop message: frame 'base_scan' at time 0.395 for reason(0)
[world_model-8] [INFO] [global_costmap.global_costmap_rclcpp_node]: [signalFailure] Drop message: frame 'base_scan' at time 0.595 for reason(0)
[world_model-8] [INFO] [global_costmap.global_costmap_rclcpp_node]: [signalFailure] Drop message: frame 'base_scan' at time 0.795 for reason(0)
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[world_model-8] [INFO] [global_costmap.global_costmap_rclcpp_node]: [signalFailure] Drop message: frame 'base_scan' at time 0.995 for reason(0)
[world_model-8] [INFO] [global_costmap.global_costmap_rclcpp_node]: [signalFailure] Drop message: frame 'base_scan' at time 1.196 for reason(0)
[dwb_controller-9] [INFO] [local_costmap.local_costmap_rclcpp_node]: [signalFailure] Drop message: frame 'base_scan' at time 1.596 for reason(0)
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [INFO] [dwb_controller]: Preempting the goal. Passing the new path to the planner.
[dwb_controller-9] [ERROR] [dwb_controller]: Resulting plan has 0 poses in it.
[dwb_controller-9] [INFO] [local_costmap.local_costmap]: Received request to clear entirely the local_costmap
[world_model-8] [INFO] [global_costmap.global_costmap]: Received request to clear entirely the global_costmap
[recoveries_node-11] [INFO] [recoveries]: Attempting Spin
[recoveries_node-11] [INFO] [recoveries]: Turning -1.57 for spin recovery.
[recoveries_node-11] [INFO] [recoveries]: Spin running...
[recoveries_node-11] [INFO] [recoveries]: Spin running...
[recoveries_node-11] [INFO] [recoveries]: Spin completed successfully
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (5.10, 0.88)
[bt_navigator-12] [ERROR] [bt_navigator]: Navigation failed
[route_manager.py-13] [INFO] [route_manager]: Result: std_msgs.msg.Empty(structure_needs_at_least_one_member=0)
[route_manager.py-13] [INFO] [route_manager]: Route mode is 'random', getting next goal
[route_manager.py-13] [INFO] [route_manager]: Sending target goal: nav2_msgs.action.NavigateToPose_Goal(pose=geometry_msgs.msg.PoseStamped(header=std_msgs.msg.Header(stamp=builtin_interfaces.msg.Time(sec=1568424822, nanosec=609596511), frame_id='map'), pose=geometry_msgs.msg.Pose(position=geometry_msgs.msg.Point(x=-6.80345344543, y=5.45276832581, z=0.0), orientation=geometry_msgs.msg.Quaternion(x=0.0, y=0.0, z=0.691578597584, w=0.722301213736))))
[bt_navigator-12] [INFO] [bt_navigator]: Begin navigating from current location to (-6.80, 5.45)
[route_manager.py-13] [INFO] [route_manager]: Goal accepted :)
[dwb_controller-9] [INFO] [dwb_controller]: Received a goal, begin following path
[dwb_controller-9] [ERROR] [GoalDistCritic]: None of the points of the global plan were in the local costmap.
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [PathDistCritic]: None of the 41 first of 41 (41) points of the global plan were in the local costmap and free
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [PathDistCritic]: None of the 41 first of 41 (41) points of the global plan were in the local costmap and free
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [GoalDistCritic]: None of the points of the global plan were in the local costmap.
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [DWBLocalPlanner]: No valid trajectories out of 398! 
[dwb_controller-9] [ERROR] [DWBLocalPlanner]: 1.00:   PathDist/Trajectory Hits Unreachable Area.
[dwb_controller-9] [ERROR] [dwb_controller]: No valid trajectories out of 398! 
[dwb_controller-9] [INFO] [local_costmap.local_costmap]: Received request to clear entirely the local_costmap
[world_model-8] [INFO] [global_costmap.global_costmap]: Received request to clear entirely the global_costmap
```
Filtering out `INFO`:
```
[bt_navigator-12] [WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
[bt_navigator-12] [ERROR] [getCurrentPose]: No Transform available Error looking up robot pose: "map" passed to lookupTransform argument target_frame does not exist. 
[bt_navigator-12] 
[bt_navigator-12] [ERROR] [getCurrentPose]: Extrapolation Error looking up robot pose: Lookup would require extrapolation into the past.  Requested time 7.53900 but the earliest data is at time 7.80200, when looking up transform from frame [base_link] to frame [map]
[bt_navigator-12] 
[bt_navigator-12] [ERROR] [getCurrentPose]: Extrapolation Error looking up robot pose: Lookup would require extrapolation into the past.  Requested time 7.57300 but the earliest data is at time 7.80200, when looking up transform from frame [base_link] to frame [map]
[bt_navigator-12] 
[dwb_controller-9] [ERROR] [dwb_controller]: Resulting plan has 0 poses in it.
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (5.10, 0.88)
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (5.10, 0.88)
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (5.10, 0.88)
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (5.10, 0.88)
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (5.10, 0.88)
[bt_navigator-12] [ERROR] [bt_navigator]: Navigation failed
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (1.73, -5.14)
[navfn_planner-10] [WARN] [navfn_planner]: Planning algorithm failed to generate a valid path to (1.73, -5.14)
[bt_navigator-12] [ERROR] [bt_navigator]: Navigation failed
[dwb_controller-9] [ERROR] [GoalDistCritic]: None of the points of the global plan were in the local costmap.
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [PathDistCritic]: None of the 41 first of 41 (41) points of the global plan were in the local costmap and free
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [PathDistCritic]: None of the 41 first of 41 (41) points of the global plan were in the local costmap and free
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [GoalDistCritic]: None of the points of the global plan were in the local costmap.
[dwb_controller-9] [WARN] [DWBLocalPlanner]: A scoring function failed to prepare
[dwb_controller-9] [ERROR] [DWBLocalPlanner]: No valid trajectories out of 398! 
[dwb_controller-9] [ERROR] [DWBLocalPlanner]: 1.00:   PathDist/Trajectory Hits Unreachable Area.
[dwb_controller-9] [ERROR] [dwb_controller]: No valid trajectories out of 398! 

```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

AAlon committed:
Use gazebo_ros launch scripts and fix route_manager

timrobotson committed:
Add ROS2 navigation

- Add navigation2 package to replace ros-planning/navigation
- Load navigation2 bringup file which loads the new move_base equivilant

AAlon committed:
Route manager conversion to ROS2

AAlon committed:
Fix route manager

AAlon commented:
https://github.com/ros-planning/navigation2/issues/1105 might be related to the first error (`No Transform available`). Not sure if it's related to the others or a red herring.

piraka9011 commented:
**System**
**OS:** Ubuntu 18.04
**Gazebo:** 9.11

I was able to get the navigation stack running with both `navigation2` and `cartographer` on the Turtlebot 3 Burger in simulation:

```bash
# In 2 separate terminals
$ ros2 launch turtlebot3_gazebo turtlebot3_world.launch.py
$ ros2 launch turtlebot3_navigation2 navigation2.launch.py map:=$HOME/map.yaml use_sim_time:=true
```

There are two things that need to be done to make sure the `navigation2` stack works:
1. `use_sim_time` should be true. Otherwise you get TF sync issues.
2. An initial pose estimate must be published passed to AMCL (`/initialpose` topic with message type `geometry_msgs/msg/PoseWithCovarianceStamped`) 

Edit:

I was able to run the route manager locally on my machine with the suggested config. I hardcoded the goals instead tho. Turtlebot moved to goal positions.

Left some comments. LGTM otherwise.

piraka9011 commented on simulation_ws/src/aws_robomaker_simulation_common/src/aws_robomaker_simulation_common/route_manager.py:
```diff
@@ -65,7 +65,9 @@ def __init__(self):
         self.route = []
 
         self.client = ActionClient(self, NavigateToPose, 'NavigateToPose')
+        time.sleep(5)
```
Don't think this is necessary since `wait_for_server()` will already wait for the client to be ready.

piraka9011 commented on simulation_ws/src/aws_robomaker_simulation_common/src/aws_robomaker_simulation_common/route_manager.py:
```diff
@@ -65,7 +65,9 @@ def __init__(self):
         self.client = ActionClient(self, NavigateToPose, 'NavigateToPose')
+        time.sleep(5)
         self.client.wait_for_server()
+        time.sleep(5)
```
ditto

AAlon commented:
> **System**
> **OS:** Ubuntu 18.04
> **Gazebo:** 9.11
> 
> I was able to get the navigation stack running with both `navigation2` and `cartographer` on the Turtlebot 3 Burger in simulation:
> 
> ```shell
> # In 2 separate terminals
> $ ros2 launch turtlebot3_gazebo turtlebot3_world.launch.py
> $ ros2 launch turtlebot3_navigation2 navigation2.launch.py map:=$HOME/map.yaml use_sim_time:=true
> ```
> 
> There are two things that need to be done to make sure the `navigation2` stack works:
> 
> 1. `use_sim_time` should be true. Otherwise you get TF sync issues.
> 2. An initial pose estimate must be published passed to AMCL (`/initialpose` topic with message type `geometry_msgs/msg/PoseWithCovarianceStamped`)
> 
> Edit:
> 
> I was able to run the route manager locally on my machine with the suggested config. I hardcoded the goals instead tho. Turtlebot moved to goal positions.
> 
> Left some comments. LGTM otherwise.

Did you run `route_manager` directly or did you launch the app with `ros2 launch cloudwatch_simulation bookstore_turtlebot_navigation.launch.py`?

I was doing the latter, and even after setting initial pose the navigation still fails with `Planning algorithm failed to generate a valid path to ...`.

Commit:
Refactor simulation_common pkg to ROS2 format

Commit:
Reformat cloudwatch_simulation pkg to ROS2 format

Commit:
Launch file for navigating with TB3

Commit:
Update gitignore

Commit:
Update gitignore

piraka9011 commented:
I've reformatted the packages to conform with the ROS2 python package format.
I've also added the launch file `turtlebot3_navigation.launch.py` to `cloudwatch_simulation` which:
1. Launches gazebo
2. Launches the navigation stack
3. Launches rviz

A params file `param/burger.yaml` with `initial_pose_x/y/a` set was also added (copypasted from `turtlebot3_navigation`

timrobotson commented on simulation_ws/src/aws_robomaker_simulation_common/aws_robomaker_simulation_common/route_manager.py:
```diff
@@ -78,48 +81,64 @@ def __init__(self):
+        self.get_logger().info('Result: {0}'.format(result.result))
+        self.route_forever()
+
+    def feedback_callback(self, feedback_msg):
```
Do you need to specify a feedback_callback if it doesn't do anything?

piraka9011 replied:
An Action will send us feedback about the long running operation (executing navigation). We should use this for debugging and/or backward compatibility.

timrobotson commented on simulation_ws/src/aws_robomaker_simulation_common/aws_robomaker_simulation_common/route_manager.py:
```diff
@@ -78,48 +81,64 @@ def __init__(self):
+
+    def get_result_callback(self, future):
+        result = future.result().result
+        # Expecting empty result (std_msgs::Empty) for NavigateToPose
```
If it keeps logging "Result: Empty" users may think this is a bug. This should be changed to something to make it clear that Empty is expected, or if it's always Empty then we don't need to print it.

piraka9011 replied:
As before, we should keep this for debugging (knowing if the server is actually working!) and backward compatibility for actions.

timrobotson commented on simulation_ws/src/cloudwatch_simulation/launch/turtlebot3_nav2.launch.py:
```diff
@@ -0,0 +1,87 @@
+        description='Full path to param file to load')
+
+    start_gazebo_cmd = IncludeLaunchDescription(
+        PythonLaunchDescriptionSource([gazebo_launch_file_dir, '/turtlebot3_world.launch.py'])
```
Should combine these using `os.path.join()`

timrobotson replied:
Also this will always load the gazebo client which we don't want when running in RoboMaker. Should load gzclient and gzserver separately and have a condition for gzclient so it only loads if `gui` param is provided like this: https://github.com/aws-robotics/aws-robomaker-sample-application-cloudwatch/blob/ros2/simulation_ws/src/cloudwatch_simulation/launch/empty_world.launch.py#L13

timrobotson commented on simulation_ws/src/cloudwatch_simulation/launch/turtlebot3_navigation.launch.py:
```diff
@@ -7,6 +7,9 @@
 
 
 def generate_launch_description():
+    TURTLEBOT3_MODEL = os.environ['TURTLEBOT3_MODEL']
```
Can get rid of this whole file now that you have turtlebot3_nav2. Or rename turtlebot3_nav2 to this.

Commit:
Update params

Commit:
Add resources

Commit:
Make Ament find the package

Commit:
Address PR comments and add more features

Commit:
Cleanup, linting, and comments

Commit:
Change callback logs to Debug

Commit:
Remove old launch file

piraka9011 commented:
The state of the current branch has everything working except for launching the nav2 stack inside  `turtlebot3_nav2.launch.py` file (works when running from command line).

Future improvement to `turtlebot3_nav2.launch.py` might be breaking it down to smaller chunks.

AAlon commented:
Thanks Anas. It's okay if we want to break it down and structure the launch scripts differently, the important thing is to maintain the same user experience of ROS1 as much as possible - specifically, you only need to invoke a single launch command, choosing either empty_world.launch.py or bookstore_turtlebot_navigation.launch.py - e.g. `ros2 launch cloudwatch_simulation bookstore_turtlebot_navigation.launch.py use_sim_time:=true`. Everything should be started without requiring additional steps.

Commit:
Add missing ROS2 pkg files, fixes warnings

Commit:
Update for debugging

timrobotson committed:
Use new nav2 launch file. Remove gazebo launching.

- Include new turtlebot3_nav2 launch file in
bookstore_turtlebot3_navigation so that it is launched with the main
application.
- Remove launching gzclient and gzserver from the turtlebot3_nav2 lunch
file as they are already launched with bookstore_world.

timrobotson committed:
Move map to cloudwatch_simulation, add waffle to params

- Move the map file from bookstore world to cloudwatch_simulation
because the path was > 128 characters.
- Create waffle and waffle_pi param files so they can be used for
navigation.
- Update map param in launch file to take the new map file correctly.

timrobotson committed:
Use ros2 branch of bookstore-world

Commit:
Fix map file and new debugging launch file

timrobotson committed:
Fix Robot Navigation

- Combined bookstore_turtlebot3_navigation and
bookstore_turtlebot_navigation together to be less confusing.
- Add initial poses to the param files so that routing works immediately
- Add 10 second sleep to route_manager so it doesn't start navigating
before the navigation stack is ready.

timrobotson commented:
The latest commit fixes navigation and setting the initial pose so the Turtlebot moves from the beginning without needing to change anything in Rviz. 

It also merges the new_btn launch file changes into bookstore_turtlebot_navigation.launch.py and uses the new navigation launch file from Anas. 

It's working correctly locally for me locally in a docker container but in robomaker all models are missing, still figuring this out.

AAlon commented:
@timrobotson can you rebase please?

timrobotson committed:
Fixing issues with launch file

timrobotson commented:
Rebased and fixed some small issues with the setup.py file for cloudwatch_common and the launch file.

AAlon commented:
Let's rerun CI once https://github.com/colcon/colcon-ros-bundle/pull/29 gets merged.

AAlon committed:
Use a version of travis-scripts which does not install colcon-bundle from source

AAlon commented on simulation_ws/src/aws_robomaker_simulation_common/setup.py:
```diff
@@ -10,13 +10,11 @@
     ],
     install_requires=[
         'setuptools',
-        'rospkg',
```
I suspect removing these breaks CI, can you check?

AAlon reviewed (commented):
We're almost there. can you also update the readme please (or open a separate PR for it)?

AAlon committed:
Update package.xml to use python3 (not 2) dependencies

AAlon committed:
Use master branch of travis-scripts

timrobotson committed:
Update README to ROS2 commands and packages

timrobotson commented:
README's have been updated to ROS2 commands and package links.

AAlon committed:
Fix readme command deploy_rotate.launch -> deploy_rotate.launch.py
